### PR TITLE
Refactor Internals

### PR DESCRIPTION
**Pull request by AydinHassan**
Goals:
 * More fine grained checks
 * Reduce code duplication
 * Facilitate the run command

Current Progress as 10/01/2015 01:50
 * Run is now available - just needs a command to hook in to it - available in another PR
 
TODO:
- [x] Build Plugin Architecture with something like an EVM. When writing a Plugin you can hook in to events during the exercise running and add results or just modify arguments and so on.
- [x] Allow for exercises to implement plugins so they can do crazy stuff themselves (like a more advanced self check) - Exercises can grab the event dispatcher from the exercise dispatcher now.
- [x] Update Database Lesson to use new Plugin architecture
- [ ] Update all Learn You PHP exercises
- [x] Move code patching to a event listener
- [x] Move composer solution installing to an event listener (PROFIT!!???)
- [x] Re-organise ExerciseDispatcher finally call
- [x] Move self-check to event
- [x] Simplify DatabaseCheck by just instantiating runner

### CODE REVIEW

**Comment by AydinHassan** [comment]
Yo @mikeymike 

This is done now. If you could cast your eye over it a little bit it would be cool, if not don't worry it is a bit big :smile:

**Comment by AydinHassan** [comment]
Now I am going to go and drink too much beer :beers:

**Comment by mikeymike** [comment]
Haha no problem dude I'll cast an eye over it tonight and tomorrow morning 

Have a good night, you deserve it ❤

File: app/config.php [hunk 1]
```diff
@@ -149,10 +166,19 @@
...
         return new HelpCommand(
             $c->get('appName'),
-            $c->get(Output::class),
+            $c->get(OutputInterface::class),
             $c->get(Color::class)
         );
     }),
+
+    //Listeners
+    PrepareSolutionListener::class      => object(),
```
Comment (mikeymike) [inline app/config.php]: Just wondering what this does?

File: src/Event/EventDispatcher.php [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            foreach ($this->listeners[$event->getName()] as $listener) {
+                $listener($event);
+            }
+        }
+
+        if (array_key_exists($event->getName(), $this->verifiers)) {
+            foreach ($this->verifiers[$event->getName()] as $verifier) {
+                $result = $verifier($event);
+
+                //return type hints pls
```
Comment (mikeymike) [inline src/Event/EventDispatcher.php]: :heart:

File: src/Event/EventDispatcher.php [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+                if ($result instanceof ResultInterface) {
+                    $this->resultAggregator->add($result);
+                } else {
+                    //??!!
```
Comment (mikeymike) [inline src/Event/EventDispatcher.php]: `InvalidResultException` ?

Comment (mikeymike) [inline src/Event/EventDispatcher.php]: Down to the dev making the workshop to ensure it's not broke

File: src/ExerciseDispatcher.php [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+     * @return ResultAggregator
+     * @throws CheckNotApplicableException
+     * @throws ExerciseNotConfiguredException
+     */
+    public function verify(ExerciseInterface $exercise, $fileName)
+    {
+        $exercise->configure($this);
+
+        $runner = $this->runnerFactory->create($exercise->getType(), $this->eventDispatcher);
+        $this->eventDispatcher->dispatch(new Event('verify.start', compact('exercise', 'fileName')));
```
Comment (mikeymike) [inline src/ExerciseDispatcher.php]: TIL `compact`

File: src/Utils/ArrayObject.php [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+    /**
+     * @param array|mixed... $array
+     */
+    public function __construct($array)
+    {
+        if (func_num_args() > 1) {
+            $this->array = func_get_args();
+        } else {
+            $this->array = $array;
```
Comment (mikeymike) [inline src/Utils/ArrayObject.php]: Should this have an elseif `is_array` check then the else be `$this->array = [$array]` ?

**Comment by mikeymike** [comment]
Some really good stuff here bro... didn't get through all of it but so far big thumbs up from me :+1:

File: app/config.php (hunk 1 above)
Comment (AydinHassan) [inline app/config.php]: Basically means just instantiate the class with no args. It is lazy so wouldnt be created until asked for, and is shared, so subsequent requests get the same instance

File: src/Event/EventDispatcher.php (hunk 3 above)
Comment (AydinHassan) [inline src/Event/EventDispatcher.php]: Yeah agreed

File: src/ExerciseDispatcher.php (hunk 4 above)
Comment (AydinHassan) [inline src/ExerciseDispatcher.php]: Same!

File: src/Utils/ArrayObject.php (hunk 5 above)
Comment (AydinHassan) [inline src/Utils/ArrayObject.php]: Yeah could do

**Pull request merged**